packages/jubmoji-api/src/proofs.ts

```diff
@@ -0,0 +1,313 @@
+  }: JubmojiInCollectionClassArgs) {
+    this.collectionPubKeys = collectionPubKeys;
+    this.sigNullifierRandomness = sigNullifierRandomness;
+    this.pathToCircuits = pathToCircuits;
+  }
+
+  async prove({
+    jubmoji,
+  }: JubmojiInCollectionProofArgs): Promise<JubmojiInCollectionProof> {
+    const { sig, msgHash, pubKey, R, T, U } =
+      getMembershipProofArgsFromJubmoji(jubmoji);
+    const index = this.collectionPubKeys.indexOf(pubKey);
+    const merkleProof = getMerkleProofFromCache(this.collectionPubKeys, index);
+    const pubKeyNullifierRandomness = getRandomNullifierRandomness();
+
+    const membershipProof = await proveMembership({
+      sig,
+      msgHash,
+      publicInputs: {
+        R,
+        T,
+        U,
+      },
+      merkleProof,
+      sigNullifierRandomness: hexToBigInt(this.sigNullifierRandomness),
+      pubKeyNullifierRandomness,
+      pathToCircuits: this.pathToCircuits,
+    });
+
+    return {
+      serializedMembershipProof: serializeMembershipProof(membershipProof),
+    };
+  }
+
+  async verify({
+    serializedMembershipProof,
+    usedSigNullifiers,
+  }: JubmojiInCollectionProof): Promise<VerificationResult> {
+    const merkleRoot = getMerkleRootFromCache(this.collectionPubKeys);
+
+    return await verifyMembership({
+      proof: deserializeMembershipProof(serializedMembershipProof),
+      merkleRoot,
+      sigNullifierRandomness: hexToBigInt(this.sigNullifierRandomness),
+      usedSigNullifiers: usedSigNullifiers?.map(hexToBigInt),
+      pathToCircuits: this.pathToCircuits,
+    });
+  }
+}
+
+// We must also pass the message nonce and random string along with the proof
+export class JubmojiInCollectionWithNonce {
+  collectionPubKeys: string[];
+  sigNullifierRandomness: string;
+  pathToCircuits?: string;
+
+  constructor({
+    collectionPubKeys,
+    sigNullifierRandomness,
+    pathToCircuits,
+  }: JubmojiInCollectionWithNonceClassArgs) {
+    this.collectionPubKeys = collectionPubKeys;
+    this.sigNullifierRandomness = sigNullifierRandomness;
+    this.pathToCircuits = pathToCircuits;
+  }
+
+  async prove({
+    jubmoji,
+  }: JubmojiInCollectionWithNonceProofArgs): Promise<JubmojiInCollectionWithNonceProof> {
+    const { sig, msgHash, pubKey, R, T, U } =
+      getMembershipProofArgsFromJubmoji(jubmoji);
+    const index = this.collectionPubKeys.indexOf(pubKey);
+    const merkleProof = getMerkleProofFromCache(this.collectionPubKeys, index);
+    const pubKeyNullifierRandomness = getRandomNullifierRandomness();
+
+    const membershipProof = await proveMembership({
+      sig,
+      msgHash,
+      publicInputs: {
+        R,
+        T,
+        U,
+      },
+      merkleProof,
+      sigNullifierRandomness: hexToBigInt(this.sigNullifierRandomness),
+      pubKeyNullifierRandomness,
+      pathToCircuits: this.pathToCircuits,
+    });
+
+    return {
+      serializedMembershipProof: serializeMembershipProof(membershipProof),
+      msgNonce: jubmoji.msgNonce,
```

Comment:
Why didn't we merge `JubmojiInCollectionProof` and `JubmojiInCollectionWithNonceProof`? And just set revealing nonce as a flag?

Reply:
Hmm, conceptually they are quite different? When we use these proofs, I think the name should be descriptive enough to say what type of proof it is. I.e. just by looking at the proof class a user of the API can know which one to use